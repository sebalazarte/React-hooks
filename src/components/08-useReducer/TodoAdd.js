import React from 'react'
import { useForm } from '../../hooks/useForm';

export const TodoAdd = ({ handleAdd}) => {

    //el useForm es usado para acceder a los valores de los input del formulario
    const [{ description }, handleInputChange, reset] = useForm({
        description: ''
    });

    const handleSubmit = (e) => {
        e.preventDefault();

        const newTodo = {
            id: new Date().getTime(),
            desc: description,
            done: false
        };

        handleAdd(newTodo);
        reset();

    }
    
    return (
        <>
            <h4>Agregar TODO</h4>
            <hr />
            <form onSubmit={handleSubmit}>
                <input
                    type="text"
                    name="description"
                    className="form-control"
                    placeholder="Aprender ..."
                    autoComplete="off"
                    onChange={handleInputChange}
                    value={description}
                    required
                />
                <button
                    className="btn btn-outline-primary btn-block mt-1"
                    type="submit">
                    Agregar
                </button>
            </form>
        </>
    )
}
