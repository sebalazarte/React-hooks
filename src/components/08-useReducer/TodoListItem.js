import React from 'react'
import PropTypes from 'prop-types';

export const TodoListItem = ({ todo, indice, handleToogle, handleDelete }) => {
    return (
        <li
            key={todo.id}
            className="list-group-item"
        >
            <p
                className={`${todo.done && 'complete'}`}
                onClick={() => handleToogle(todo.id)}
            >
                {indice + 1}. {todo.desc}
            </p>
            <button
                className="btn btn-outline-danger"
                type="button"
                onClick={() => handleDelete(todo.id)}
            >
                Borrar
            </button>
        </li>
    )
}

TodoListItem.propTypes = {
    todo: PropTypes.any.isRequired,
    indice: PropTypes.number.isRequired,
    handleToogle: PropTypes.func.isRequired,
    handleDelete: PropTypes.func.isRequired
}