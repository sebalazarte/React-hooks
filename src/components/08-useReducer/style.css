body{
    padding: 50px;
}

.list-group-item{
    align-items: center;
    display: flex;
    justify-content: space-between;
}
p{
    margin: 0px;
    cursor: pointer;
}

.complete{
    text-decoration: line-through;
}