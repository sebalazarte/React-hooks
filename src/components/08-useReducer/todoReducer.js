export const todoRedurer = (state = [], action) => {
    switch (action.type) {
        case 'add':
            console.log(action.payload);
            return [...state, action.payload];
        case 'delete':
            return state.filter(todo => todo.id !== action.payload);
        case 'toogle-old':
            return state.map(todo => {
                if (todo.id === action.payload) {
                    return {
                        ...todo,
                        done: !todo.done
                    }
                } else {
                    return todo;
                }
            })
        case 'toogle':
            return state.map(todo => 
                (todo.id === action.payload)
                    ? {...todo, done: !todo.done}
                    : todo
                );
        default:
            return state;
    }
}