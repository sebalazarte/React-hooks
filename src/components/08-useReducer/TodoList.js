import React from 'react'
import PropTypes from 'prop-types';
import { TodoListItem } from './TodoListItem';

export const TodoList = ({ todos = [], handleToogle, handleDelete }) => {
    return (
        <ul className="list-group list-group-flush">
            {
                todos.map((todo, i) => (
                    <TodoListItem
                        todo={todo}
                        indice={i}
                        handleToogle={handleToogle}
                        handleDelete={handleDelete}
                    />
                ))
            }
        </ul>
    )
}

TodoList.propTypes = {
    todos: PropTypes.array,
    handleToogle: PropTypes.func.isRequired,
    handleDelete: PropTypes.func.isRequired
}

