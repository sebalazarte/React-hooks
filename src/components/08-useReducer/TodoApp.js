import React, { useReducer, useEffect } from 'react'
import './style.css';
import { todoRedurer } from './todoReducer';
import { useForm } from '../../hooks/useForm';

// const initialState = [{
//     id: new Date().getTime(),
//     desc: 'Aprender React',
//     done: false
// }];

const init = () => {
    return JSON.parse(localStorage.getItem('todos')) || [];
}

export const TodoApp = () => {

    // const [todos, dispatch] = useReducer(todoRedurer, initialState, init);
    //Al useRedurer se le puede pasar un estado incial o una funcion que carga el estado inicial
    //en este caso la funcion init trae el estado del localstorage
    const [todos, dispatch] = useReducer(todoRedurer, [], init);

    //el useForm es usado para acceder a los valores de los input del formulario
    const [{description}, handleInputChange, reset] = useForm({
        description: ''
    });

    //el efecto guarda en el localstorage cualquier cambio que tengan los todos
    useEffect(() => {
        localStorage.setItem('todos', JSON.stringify(todos));
    }, [todos])

    const handleSubmit = (e) => {
        e.preventDefault();

        const newTodo = {
            id: new Date().getTime(),
            desc: description,
            done: false
        };

        const action = {
            type: 'add',
            payload: newTodo
        }

        dispatch(action);
        reset();
    }

    return (
        <div>

            <h1>Todo App ({todos.length})</h1>
            <hr />
            <div className="row">
                <div className="col-7">
                    <ul className="list-group list-group-flush">
                        {
                            todos.map((todo, i) => (
                                <li
                                    key={todo.id}
                                    className="list-group-item"
                                >
                                    <p className="text-center">{i + 1}. {todo.desc}</p>
                                    <button className="btn btn-outline-danger" type="button">Borrar</button>
                                </li>
                            ))
                        }
                    </ul>
                </div>
                <div className="col-5">
                    <h4>Agregar TODO</h4>
                    <hr />
                    <form onSubmit={handleSubmit}>
                        <input
                            type="text"
                            name="description"
                            className="form-control"
                            placeholder="Aprender ..."
                            autoComplete="off"
                            onChange={handleInputChange}
                            value={description}
                            required
                        />
                        <button
                            className="btn btn-outline-primary btn-block mt-1"
                            type="submit">
                            Agregar
                        </button>
                    </form>

                </div>
            </div>
        </div>
    )
}


