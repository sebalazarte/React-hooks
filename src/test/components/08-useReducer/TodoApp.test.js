import React from 'react';
import { act } from '@testing-library/react'
import { TodoApp } from '../../../components/08-useReducer/TodoApp';
import { demoTodos } from '../../fixtures/demoTodos';
const { shallow, mount } = require("enzyme");

describe('TodoApp', () => {

    const wrapper = shallow(<TodoApp />);

    Storage.prototype.setItem = jest.fn(()=>{});

    test('debe mostrarse correctamente', () => {
        expect(wrapper).toMatchSnapshot();
    });

    test('debe agregar un todo', () => {

        const wrapper = mount(<TodoApp />);
        act(() => {
            wrapper.find('TodoAdd').prop('handleAdd')(demoTodos[0]);
            wrapper.find('TodoAdd').prop('handleAdd')(demoTodos[1]);
        });

        expect(wrapper.find('h1').text().trim()).toBe('Todo App (2)');
        expect(localStorage.setItem).toBeCalledTimes(2);
    });

    test('debe eliminar un todo', () => {
        
        wrapper.find('TodoAdd').prop('handleAdd')(demoTodos[0]);        
        wrapper.find('TodoList').prop('handleDelete')(demoTodos[0].id);
        expect(wrapper.find('h1').text().trim()).toBe('Todo App (0)');
    });

});