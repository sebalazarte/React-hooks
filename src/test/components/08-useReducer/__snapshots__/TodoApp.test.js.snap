// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TodoApp debe mostrarse correctamente 1`] = `
<div>
  <h1>
    Todo App (
    0
    )
  </h1>
  <hr />
  <div
    className="row"
  >
    <div
      className="col-7"
    >
      <TodoList
        handleDelete={[Function]}
        handleToogle={[Function]}
        todos={Array []}
      />
    </div>
    <div
      className="col-5"
    >
      <TodoAdd
        handleAdd={[Function]}
      />
    </div>
  </div>
</div>
`;
